### PR TITLE
Refactor ExperimentModel to only export methods

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes

This PR refactors the `ExperimentModel` so we are no longer exporting the model itself and mimics the pattern of the `FeatureModel` by introducing a `toInterface()` method.

### Testing

- [x] Confirm the “New Custom Report works as expected
- [x] Confirm getExperimentById method returns the correct experiment
     - [ ] When clicking "Download Notebook" via an experiment, the experiment returned was correct 
     **Currently getting a 400 error - No module named 'nbformat'. This is happening on main as well**
- [x] Confirm the getAllExperiments returns all of the experiments as expected - returned 13 experiments correctly
    - [x] Also confirmed the filter by project worked as expected
- [x] Confirm the getExperimentByTrackingKey method returns the expected experiment - e.g. http://localhost:3000/experiments/lookup?trackingKey=test12345678 redirects correctly to http://localhost:3000/experiment/exp_exl5j1bt3ld366pou
- [x] Confirm that getExperimentsByIds returns experiments included in the array passed in to the method - e.g. 
    - [x] getPresentation returns the presentation correctly
    - [x] getPresentationPreview correctly returns the correct experiments based on inputs from the new presentations first screen. Ex: If you create a report with 3 experiments, those 3 experiments are correctly referenced in the report. If you remove 1, that is then reflected in the presentation
- [x] deleteExperimentById correctly deletes a single experiment
- [x] getExperimentsUsingSegments returns the correct experiment - when deleting a segment used for an experiment, we’re correctly warning the user that it’s connected to an experiment, and if it’s not connected to any experiments, we don’t show the message correctly
- [x] Ensure getSampleExperiment returns the correct sample experiment
    - [ ] Currently having trouble testing, I found a way to click the Import Sample Data button via ExperimentsGetStarted, but got an error - need to revisit
- [x] Ensure createExperiment creates a new experiment as expected
    - [ ] I can’t figure out how to test the createExperiment call from adminController.addSampleData - I get a permission error when trying to view.
- [x] Ensure that updateExperimentById works as expected
    - [x] I tested archiving/unarchiving an experiment and it worked as expected
    - [x] I tested updating the status of an experiment and it worked as expected
    - [x] I tested stopping the experiment via the stop button on the old version of the experiment page and it worked as expected
    - [x] I tested deleting experiment phase and that worked as expected
    - [x] I tested creating a new experiment phase and that worked as expected
    - [x] I tested updating an experiment phase and that worked as expected
    - [x] I tested adding a screenshot and that worked as expected
    - [x] I tested removing a screenshot and that worked as expected
- [x] Ensure getExperimentsByMetric works as expected - when deleting a metric, I can confirm that the experiment connected to a metric was returned correctly.
- [x] Ensure removeMetricFromExperiment correctly removes a metric from an experiment
- [x] Ensure removeProjectFromExperiments correctly removes a project from a experiment.project
- [x] Ensure getExperimentByIdea returns the correct experiment
- [x] Ensure the getExperimentsByQuery correctly returns the experiment
- [x] Ensure the updateExperimentByQuery correctly updates the experiment - I tested deleting a segment connected to an experiment and can confirm the segment was removed from the experiment.

### CODE REVIEW

File: packages/back-end/src/controllers/metrics.ts [hunk 1]
```diff
@@ -253,7 +255,7 @@ export async function getMetric(
     });
   }
 
-  const experiments = await ExperimentModel.find(
+  const experiments = await getExperimentsByQuery(
```
Comment (jdorn) [inline packages/back-end/src/controllers/metrics.ts:258]: I'd like to avoid having Mongo queries outside of the Model files.  Here, for example, it's getting the most recent experiments that are using a specific metric and it's only returning summary fields instead of the full interface.  We could create a dedicated function in the Model file for this:

```ts
// Type with only the subset of fields being selected
export type MetricExperimentSummary = Pick<ExperimentInterface, "id", "name", "status", "phases", "results", "analysis">;
export function getRecentExperimentsUsingMetric(metric: MetricInterface): MetricExperimentSummary[] {
  ...
}
```

File: packages/back-end/src/controllers/reports.ts [hunk 2]
```diff
@@ -34,10 +36,7 @@ export async function postReportFromSnapshot(
     throw new Error("Invalid snapshot id");
   }
 
-  const experiment = await ExperimentModel.findOne({
-    organization: org.id,
-    id: snapshot.experiment,
-  });
+  const experiment = await getExperimentById(snapshot.experiment);
```
Comment (jdorn) [inline packages/back-end/src/controllers/reports.ts:39]: Experiment ids are not unique across organizations.  So the org id always has to be part of the query

File: packages/back-end/src/jobs/updateExperimentResults.ts [hunk 3]
```diff
@@ -39,7 +43,7 @@ export default async function (agenda: Agenda) {
 
     // New way, based on dynamic schedules
     const experimentIds = (
-      await ExperimentModel.find(
+      await getExperimentsByQuery(
```
Comment (jdorn) [inline packages/back-end/src/jobs/updateExperimentResults.ts:46]: Same here - create a dedicated model function instead of having a Mongo query here

File: packages/back-end/src/jobs/updateExperimentResults.ts [hunk 4]
```diff
@@ -237,8 +239,8 @@ async function updateSingleExperiment(job: UpdateSingleExpJob) {
     // If we failed to update the experiment, turn off auto-updating for the future
     try {
       experiment.autoSnapshots = false;
-      experiment.markModified("autoSnapshots");
-      await experiment.save();
+      changes.autoSnapshots = false;
+      await updateExperimentById(experiment, changes);
```
Comment (jdorn) [inline packages/back-end/src/jobs/updateExperimentResults.ts:243]: In this case, you can remove the `changes` variable entirely and just make this call inline:

```ts
await updateExperimentById(experiment, { autoSnapshots: false })
```

The way the code is currently, it implies that somewhere else in the ~100 lines above is also adding changes to the experiment, but that's not the case here.

File: packages/back-end/src/models/ExperimentModel.ts [hunk 5]
```diff
@@ -98,7 +106,292 @@ const experimentSchema = new mongoose.Schema({
...
+/**
+ * Convert the Mongo document to an ExperimentInterface, omitting Mongo default fields __v, _id
+ * @param doc
+ */
+const toInterface = (doc: ExperimentDocument): ExperimentInterface =>
+  omit(doc.toJSON(), ["__v", "_id"]);
+
+export async function getExperimentById(
+  id: string
+): Promise<ExperimentInterface | null> {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:125]: The experiment id is not unique across organizations, so this needs to include the org id in the query

File: packages/back-end/src/models/ExperimentModel.ts [hunk 6]
```diff
@@ -98,7 +106,292 @@ const experimentSchema = new mongoose.Schema({
...
+    organization,
+    trackingKey,
+  });
+
+  return experiment ? toInterface(experiment) : null;
+}
+
+export async function getExperimentsByIds(
+  ids: string[]
+): Promise<ExperimentInterface[]> {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:161]: Same here.  Query must also include the org id

File: packages/back-end/src/models/ExperimentModel.ts [hunk 7]
```diff
@@ -98,7 +106,292 @@ const experimentSchema = new mongoose.Schema({
...
+  const experiments = await ExperimentModel.find({
+    id: { $in: ids },
+  });
+
+  return experiments.map((m) => toInterface(m));
+}
+
+export async function deleteExperimentById(id: string): Promise<void> {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:169]: Same - query must include org id

File: packages/back-end/src/models/ExperimentModel.ts [hunk 8]
```diff
@@ -98,7 +106,292 @@ const experimentSchema = new mongoose.Schema({
...
+export async function createExperiment(
+  data: Partial<ExperimentInterface>,
+  organization: OrganizationInterface
+): Promise<ExperimentInterface> {
+  if (!data.organization) {
+    throw new Error("Missing organization");
+  }
+  if (data.organization !== organization.id) {
+    throw new Error("Experiment and Organization must match");
+  }
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:210]: Instead of this, we could just set `data.organization = organization.id` to force it to always match.  That way, the calling function doesn't even need to specify `experiment.organization`, it will be added for them.

File: packages/back-end/src/models/ReportModel.ts [hunk 9]
```diff
@@ -61,12 +61,13 @@ export async function getReportsByOrg(
...
-    const allExperiments = await getExperimentsByOrganization(
-      organization,
-      project
-    );
+    const allExperiments = await getAllExperiments(organization, project);
     const expIds = new Set(allExperiments.map((e) => e.id));
-    reports = reports.filter((r) => expIds.has(r.experimentId));
+    reports = reports.filter((r) => {
+      if (r.experimentId) {
+        expIds.has(r.experimentId);
```
Comment (jdorn) [inline packages/back-end/src/models/ReportModel.ts:68]: Filter functions need to return true or false, but this is always returning null.  Before, it wasn't using curly braces around the function body, which means it automatically returns.

An easy fix is to remove the curly braces again

```ts
reports = reports.filter((r) => r.experimentId && expIds.has(r.experimentId));
```

Or, add return statements to your new code:

```ts
reports = reports.filter((r) => {
  if (r.experimentId) {
    return expIds.has(r.experimentId);
  }
  return false;
});
```

**Review by jdorn — CHANGES_REQUESTED** [review]

File: packages/back-end/src/models/ReportModel.ts (hunk 9 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ReportModel.ts:68]: DOH! Thanks

File: packages/back-end/src/models/ExperimentModel.ts [hunk 10]
```diff
@@ -355,43 +360,178 @@ export async function removeProjectFromExperiments(
...
+    }
+  )
+    .limit(100)
+    .sort({ nextSnapShotAttempt: 1 });
+  return experiments.map((m) => toInterface(m));
+}
+
+export async function getExperimentsToUpdateLegacy(
+  latestDate: Date
+): Promise<Pick<ExperimentInterface, "id" | "organization">[]> {
```
Comment (mknowlton89) [inline packages/back-end/src/models/ExperimentModel.ts:405]: @jdorn For all of these new methods that replaced the general `getExperimentsByQuery` method, I opted to define the return type in-line. It does get a bit long with the `Pick` statement - let me know if you'd prefer I define these above and reference here.

File: packages/back-end/types/experiment.d.ts [hunk 11]
```diff
@@ -115,3 +115,20 @@ export type ExperimentInterfaceStringDates = Omit<
   dateUpdated: string;
   phases: ExperimentPhaseStringDates[];
 };
+
+export type ChangeLog = {
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/back-end/types/experiment.d.ts]: You bet - just updated!

File: packages/back-end/src/models/ExperimentModel.ts [hunk 12]
```diff
@@ -107,18 +121,377 @@ const experimentSchema = new mongoose.Schema({
   ideaSource: String,
 });
 
-export const ExperimentModel = mongoose.model<ExperimentDocument>(
+type ExperimentDocument = mongoose.Document & ExperimentInterface;
```
Comment (tinahollygb) [inline packages/back-end/src/models/ExperimentModel.ts:124]: 🙌

File: packages/back-end/src/models/ExperimentModel.ts [hunk 13]
```diff
@@ -107,18 +121,377 @@ const experimentSchema = new mongoose.Schema({
...
+): Promise<ExperimentInterface[]> {
+  const query: FilterQuery<ExperimentDocument> = {
+    organization,
+  };
+
+  if (project) {
+    query.project = project;
+  }
+
+  return (await ExperimentModel.find(query)).map((m) => toInterface(m));
```
Comment (tinahollygb) [inline packages/back-end/src/models/ExperimentModel.ts:158]: This should work too since it has the same signature:

```suggestion
  return (await ExperimentModel.find(query)).map(toInterface);
```

Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:158]: Woah! I had no idea this shorthand existed. Very cool.

File: packages/back-end/src/controllers/experiments.ts [hunk 14]
```diff
@@ -943,41 +954,41 @@ export async function postExperimentStop(
   }
   req.checkPermissions("createAnalyses", exp.project);
 
-  const existing = exp.toJSON();
+  const existing = exp;
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/back-end/src/controllers/experiments.ts:957]: Ahh... that's a good point.

As we've now introduced the `toInterface` method within the model, we loose access to Mongo's `toJSON()` method. (Atleast, thats what I think is happening).

I'll move forward with using `cloneDeep`. I just ran through all of the methods where we were doing `const existing = exp;` and changed it to `const existing = cloneDeep(exp);` and all of the methods worked as expected.

Thanks!

File: packages/back-end/src/models/ExperimentModel.ts [hunk 15]
```diff
@@ -107,18 +121,377 @@ const experimentSchema = new mongoose.Schema({
...
+      },
+    },
+    {
+      id: true,
+      organization: true,
+    }
+  )
+    .limit(100)
+    .sort({ nextSnapShotAttempt: 1 });
+  return experiments.map(toInterface);
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:357]: The toInterface method is meant to deal with entire experiment objects, but this query and a few below it only select a subset of fields.

File: packages/back-end/src/models/ExperimentModel.ts [hunk 16]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+    });
+    if (previous) {
+      await logExperimentDeleted(organization, previous);
+    }
+  } catch (e) {
+    logger.error(e);
+  }
+
+  await ExperimentModel.deleteOne({
+    id,
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:588]: This needs to have an org Id in the query

File: packages/back-end/src/models/ExperimentModel.ts [hunk 17]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+ */
+export async function deleteExperimentByIdForOrganization(
+  id: string,
+  organization: OrganizationInterface
+) {
+  try {
+    const previous = await findExperiment({
+      experimentId: id,
+      organizationId: organization.id,
+    });
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:579]: I’m pretty sure the experiment controller is already fetching the experiment from Mongo. We should just pas that in here instead of the id.

File: packages/back-end/src/models/ExperimentModel.ts [hunk 18]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+      previous,
+      current,
+    });
+  });
+};
+
+export async function getExperimentsByOrganization(
+  organization: string,
+  project?: string
+) {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:627]: Is this different from getAllExperiments?

File: packages/back-end/src/models/ExperimentModel.ts [hunk 19]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+  tag: string;
+}): Promise<void> => {
+  const query = { organization: organization.id, tags: tag };
+  const previousExperiments = await ExperimentModel.find(query);
+
+  await ExperimentModel.updateMany(query, {
+    $pull: { tags: tag },
+  });
+
+  previousExperiments.forEach((previous) => {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:612]: Previous needs to run through the toInterface method

File: packages/back-end/src/models/ExperimentModel.ts [hunk 20]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+export async function deleteExperimentByIdForOrganization(
+  id: string,
+  organization: OrganizationInterface
+) {
+  try {
+    const previous = await findExperiment({
+      experimentId: id,
+      organizationId: organization.id,
+    });
+    if (previous) {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:580]: Previous needs to run through the toInterface method

File: packages/back-end/src/models/ExperimentModel.ts [hunk 21]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+  const activationMetricQuery = {
+    organization: orgId,
+    activationMetric: metricId,
+  };
+  const docsToTrackChanges = await ExperimentModel.find({
+    $or: [metricQuery, guardRailsQuery, activationMetricQuery],
+  });
+  docsToTrackChanges.forEach((experiment: ExperimentDocument) => {
+    if (!oldExperiments[experiment.id]) {
+      oldExperiments[experiment.id] = {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:666]: Need to use toInterface on experiment

File: packages/back-end/src/models/ExperimentModel.ts [hunk 22]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+    $pull: { guardrails: metricId },
+  });
+
+  // Remove from activationMetric
+  await ExperimentModel.updateMany(activationMetricQuery, {
+    $set: { activationMetric: "" },
+  });
+
+  const ids = Object.keys(oldExperiments);
+  const updatedExperiments = await ExperimentModel.find({ id: { $in: ids } });
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:689]: Need to use org Id in the query and use toInterface on the result. Also, we might not even need this query. We already have all the experiments in memory and know how they changed. We can use the same cloneDeep method we use elsewhere

File: packages/back-end/src/models/ExperimentModel.ts [hunk 23]
```diff
@@ -190,6 +563,181 @@ export const logExperimentUpdated = async ({
...
+export async function removeProjectFromExperiments(
+  project: string,
+  organization: OrganizationInterface
+) {
+  const query = { organization: organization.id, project };
+  const previousExperiments = await ExperimentModel.find(query);
+
+  await ExperimentModel.updateMany(query, { $set: { project: "" } });
+
+  previousExperiments.forEach((previous) => {
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:720]: Need to use toInterface here

**Review by jdorn — DISMISSED** [review]

File: packages/back-end/src/models/ExperimentModel.ts [hunk 24]
```diff
@@ -107,18 +121,377 @@ const experimentSchema = new mongoose.Schema({
...
+
+  return exp ? toInterface(exp) : null;
+}
+
+export async function createExperiment(
+  data: Partial<ExperimentInterface>,
+  organization: OrganizationInterface
+): Promise<ExperimentInterface> {
+  if (!data.organization) {
+    throw new Error("Missing organization");
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:201]: I don't *think* so, now that we've updated the line below this.

File: packages/back-end/src/models/ExperimentModel.ts (hunk 15 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:357]: Updated to simply `return experiments`

File: packages/back-end/src/models/ExperimentModel.ts (hunk 17 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:579]: If we pass the experiment in (instead of the id) can't we just remove this call to find the `previous` and just call `await logExperimentDeleted(organization, experiment)` before calling `await ExperimentModel.deleteOne...`?

I'm a little confused about the `previous` nomenclature here, as to me it indicates we're trying to find maybe the experiment other than the one we're trying to delete, which, looking at the `logExperimentDeleted` that is just logging which experiment we're about to delete. I also dropped the await ExperimentModel.deleteOne call into the try/catch block.

Let me know what you think of this revised flow.

```
/**
 * Deletes an experiment by ID and logs the event for the organization
 * @param id
 * @param organization
 */
export async function deleteExperimentByIdForOrganization(
  experiment: ExperimentInterface,
  organization: OrganizationInterface
) {
  try {
    await logExperimentDeleted(organization, experiment);

    await ExperimentModel.deleteOne({
      id: experiment.id,
      organization: organization.id,
    });
  } catch (e) {
    logger.error(e);
  }
}
```

File: packages/back-end/src/models/ExperimentModel.ts (hunk 18 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:627]: No - I removed it, this was an artifact that I didn't catch when merging Tina's changes. It wasn't being called from anywhere.

File: packages/back-end/src/models/ExperimentModel.ts (hunk 20 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:580]: Now that we've updated this to take in the experiment that is returned from `getExperimentById` rather than getting it again, it should now already be the ExperimentInterface rather than the ExperimentDocument.

File: packages/back-end/src/models/ExperimentModel.ts (hunk 23 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:720]: Updated this to:

```
export async function removeProjectFromExperiments(
  project: string,
  organization: OrganizationInterface
) {
  const query = { organization: organization.id, project };
  const previousExperiments = await ExperimentModel.find(query);

  await ExperimentModel.updateMany(query, { $set: { project: "" } });

  const previousExperimentsInterface = previousExperiments.map(toInterface);

  previousExperimentsInterface.forEach((previous) => {
    const current = cloneDeep(previous);
    current.project = "";

    logExperimentUpdated({
      organization,
      previous,
      current,
    });
  });
}
```

I feel like there is a better pattern than creating the `previousExperimentsInterface` array only to loop through it.

The alternative of chaining previousExperiments.map(toInterface).forEach((previous) =>.....` feels very verbose, and a bit hard to follow, though.

Any suggestions?

File: packages/back-end/src/models/ExperimentModel.ts (hunk 22 above)
Comment (mknowlton89) [reply packages/back-end/src/models/ExperimentModel.ts:689]: I've updated the query for now - at our 1:1 today let's talk through the alternate approach of using `cloneDeep`.

File: packages/back-end/src/controllers/experiments.ts [hunk 25]
```diff
@@ -701,55 +692,48 @@ export async function postExperiment(
...
-      id: exp.id,
+      id: experiment.id,
     },
-    details: auditDetailsUpdate(existing, exp),
+    details: auditDetailsUpdate(experiment, updated),
   });
 
   // If there are new tags to add
-  await addTagsDiff(org.id, existing.tags || [], data.tags || []);
+  await addTagsDiff(org.id, updated?.tags || [], data.tags || []);
```
Comment (jdorn) [inline packages/back-end/src/controllers/experiments.ts:730]: This should be `experiment`, not `updated`.  You want to pass both the old tags and the new ones in.  You are passing the new tags twice.

File: packages/back-end/src/controllers/experiments.ts [hunk 26]
```diff
@@ -1016,41 +988,39 @@ export async function deleteExperimentPhase(
...
+  if (phaseIndex < 0 || phaseIndex >= experiment.phases?.length) {
     throw new Error("Invalid phase id");
   }
 
-  const existing = cloneDeep(exp);
-
   // Remove phase from experiment and revert to draft if no more phases left
-  exp.phases.splice(phaseIndex, 1);
-  changes.phases = exp.phases;
+  experiment.phases.splice(phaseIndex, 1);
```
Comment (jdorn) [inline packages/back-end/src/controllers/experiments.ts:1017]: `splice` mutates the original array, which is not what you want here.  Instead, you can just use `filter`.

```ts
// Remove an element from an array without mutating the original
changes.phases = experiment.phases.filter((phase, i) => i !== phaseIndex);
```

File: packages/back-end/src/controllers/experiments.ts [hunk 27]
```diff
@@ -1016,41 +988,39 @@ export async function deleteExperimentPhase(
...
+  changes.phases = experiment.phases;
 
-  if (!exp.phases.length) {
+  if (!experiment.phases.length) {
```
Comment (jdorn) [inline packages/back-end/src/controllers/experiments.ts:1020]: You'll want to check `changes.phases.length` here since you want to get the length after deleting the phase.

File: packages/back-end/src/controllers/experiments.ts [hunk 28]
```diff
@@ -1641,49 +1603,49 @@ export async function addScreenshot(
...
     });
     return;
   }
 
-  exp.variations[variation].screenshots =
-    exp.variations[variation].screenshots || [];
-  exp.variations[variation].screenshots.push({
+  experiment.variations[variation].screenshots =
+    experiment.variations[variation].screenshots || [];
+  experiment.variations[variation].screenshots.push({
```
Comment (jdorn) [inline packages/back-end/src/controllers/experiments.ts:1635]: This is mutating the `experiment` variable.  Instead, you want to leave the original `experiment` alone and only modify the `changes` variable.

File: packages/back-end/src/models/ExperimentModel.ts [hunk 29]
```diff
@@ -286,6 +287,14 @@ export async function updateExperimentById(
       $set: changes,
     }
   );
+
+  const updated = await getExperimentById(organization, experiment.id);
```
Comment (jdorn) [inline packages/back-end/src/models/ExperimentModel.ts:291]: We don't need to re-fetch from the database.  You have the original `experiment` object and `changes`.  So you can just do 

```ts
const updated = {
  ...experiment,
  ...changes
}
```

File: packages/back-end/src/controllers/experiments.ts [hunk 30]
```diff
@@ -1632,13 +1631,14 @@ export async function addScreenshot(
 
   experiment.variations[variation].screenshots =
     experiment.variations[variation].screenshots || [];
-  experiment.variations[variation].screenshots.push({
+
+  changes.variations = experiment.variations;
+
+  changes.variations[variation].screenshots.push({
```
Comment (jdorn) [inline packages/back-end/src/controllers/experiments.ts:1637]: This will still mutate the original.  You need to deep clone `experiment.variations` first before modifying it.

Comment (mknowlton89) [reply packages/back-end/src/controllers/experiments.ts:1638]: @jdorn Got this updated. Let's talk through this at our 1:1 today as I don't fully grok why the `deepClone` is necessary here as we don't appear to be using the original `experiment` object (other than it's `id`) or returning it later in this method.

Additionally, when looking at the `deleteScreenshot` method above, it looks like we are doing something similar, but not using cloneDeep there - we're directly mutating the `experiment` object via the use of the `.filter` method, before saving `changes.variations = experiment.variations`. Should we also be leveraging `cloneDeep` there?

If not, let's talk through why.

**Review by jdorn — APPROVED** [review]
